proteus/LinearAlgebraTools.py

```diff
@@ -724,8 +983,166 @@ def apply(self,A,x,y):
         -------
         y : vector
         """
+#        _petsc_view(x,'x_mpi_2')
         self.ksp.solve(x,y)
+#        _petsc_view(y,'y_mpi_2')
+#        _petsc_view(self.ksp.getOperators()[0],'Qp_mpi_2')
+#        import pdb ; pdb.set_trace()
+
+class Sp_shell(InvOperatorShell):
+    def __init__(self, A00, A11, A01, A10):
+        self.A00 = A00
+        self.A11 = A11
+        self.A01 = A01
+        self.A10 = A10
+        self._create_Sp()
+        self._options = p4pyPETSc.Options()
+        self._create_constant_nullspace()
+        
+        self.kspSp = p4pyPETSc.KSP().create()
+        self.kspSp.setOperators(self.Sp,self.Sp)
+        self.kspSp.setOptionsPrefix('innerSp_solve_')
+        self.kspSp.setFromOptions()
+        self.Sp.setNullSpace(self.const_null_space)
+        self.kspSp.setUp()
+
+    def apply(self,A,x,y):
+        tmp1 = p4pyPETSc.Vec().create()
+        tmp1 = x.copy()
+        self.const_null_space.remove(tmp1)
+        self.kspSp.solve(tmp1,y)
+
+    def _create_Sp(self):
+        self.A00_inv = petsc_create_diagonal_inv_matrix(self.A00)
+        A00_invBt = self.A00_inv.matMult(self.A01)
+        self.Sp = self.A10.matMult(A00_invBt)
+#        self.Sp.scale(-1.)
+        self.Sp.aypx(-1.,self.A11)
+    
+
+class TwoPhase_PCDInv_shell(InvOperatorShell):
+    r""" Shell class for the two-phase PCD preconditioner.  The
```

Comment:
Thanks for putting the time into the doc string. It's good to use a consistent notation for these preconditioners.